### PR TITLE
Provide capacity of block-mode RBD volumes

### PR DESCRIPTION
**Pull request by nixpanic**
# Describe what this PR does #

Metrics from block-mode volumes are not available from Kubelet. CephFS and RBD-with-filesystem do return metrics. This adds support for the `capacity` attribute for block-mode volumes.

## Related issues ##

It seems Kubernetes does not call `NodeGetVolumeStats` got block-mode volumes yet. This will get added through https://github.com/kubernetes/kubernetes/pull/97972 _(merged)_

Without the change in Kubernetes, this can not be tested (and we do not have tests for metrics either yet).

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

File: e2e/pvc.go [hunk 1]
```diff
@@ -235,3 +236,43 @@ func checkPVSelectorValuesForPVC(f *framework.Framework, pvc *v1.PersistentVolum
...
 }
+
+func getMetricsForPVC(f *framework.Framework, pvc *v1.PersistentVolumeClaim, t int) error {
+	kubelet, err := getKubeletIP(f.ClientSet)
+	if err != nil {
+		return err
+	}
+
+	// kubelet needs to be started with --read-only-port=10255
+	cmd := fmt.Sprintf("curl --silent 'http://%s:10255/metrics'", kubelet)
```
Comment (mykaul) [inline e2e/pvc.go:256]: Nits:
1. Why not use 127.0.0.1? (so we don't need DNS, rely on network)?
2. '10255' is the default. Do we foresee it'll ever change? (https://github.com/Azure/AKS/issues/1601 ? https://github.com/kubernetes/kubeadm/issues/732 ?)

Comment (nixpanic) [reply e2e/pvc.go:256]: 1. 127.0.0.1 will be the container (Rook Ceph Toolbox), not an address where Kubelet listens on
2. port 10255 used to be the default, it is now disabled unless `--read-only-port` is passed to Kubelet (which we do with the `scripts/minikube.sh`, once that works)

As an alternative for (2), we could add additional permissions to some Pod where we can get the metrics from (the node may not be reachable from the system where the e2e tests got started, hence the toolbox pod). Not all pods have permission to read the non-readonly port, and that needs certificates to authenticate.

Comment (Madhu-1) [reply e2e/pvc.go:256]: it might be possible to get the metrics from the E2E framework, can we please check Is there any helper functions already available for this one at https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/metrics

Comment (nixpanic) [reply e2e/pvc.go:256]: That looks like an option, I have added a commit that rewrites the test to use that.

Comment (nixpanic) [reply e2e/pvc.go:256]: When using the metrics gathering from the Kubernetes E2E framework, I do get
```
kubelet_container_log_filesystem_used_bytes{container="rhel8", namespace="block-stats", pod="rhel8-rbd-block", uid="87d26b70-a1c4-4017-a29b-9336d9a187c5"} => 12288
```
This does not call the CSI NodeGetVolumeStats procedure, so it is not useful for this testing, unfortunately.

File: scripts/minikube.sh [hunk 2]
```diff
@@ -175,7 +175,7 @@ fi
 K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-"BlockVolume=true,CSIBlockVolume=true,VolumeSnapshotDataSource=true,ExpandCSIVolumes=true"}
 
 #extra-config for kube https://minikube.sigs.k8s.io/docs/reference/configuration/kubernetes/
-EXTRA_CONFIG=${EXTRA_CONFIG:-"--extra-config=apiserver.enable-admission-plugins=PodSecurityPolicy"}
+EXTRA_CONFIG_PSP="--extra-config=apiserver.enable-admission-plugins=PodSecurityPolicy"
```
Comment (pkalever) [inline scripts/minikube.sh:168]: @nixpanic Will it be ideal to split the changes to minikube.sh into a separate patch ?

Comment (nixpanic) [reply scripts/minikube.sh:168]: The change here is really needed for the e2e testing as the `--read-only-port` option is required to be passed at the 1st (and maybe subsequent) `minikube start` commands, that conflicts with the PSP option. I do not think there is much benefit in adding it as a separate commit.

Comment (pkalever) [reply scripts/minikube.sh:168]: I just asked for this as the context was not clear for me.

**Comment by nixpanic** [comment]
kubernetes/kubernetes#97972 has been merged for Kubernetes 1.22. This should now be ready for merging. e2e tests will be skipped when Kubernetes < 1.22 is used.

**Review by humblec — DISMISSED** [review]

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e-helm/k8s-1.19

Rook operator failed to start:
```
failed to get pod: Get https://10.96.0.1:443/api/v1/namespaces/rook-ceph/pods/rook-ceph-operator-864f5d5868-5795c: dial tcp 10.96.0.1:443: connect: network is unreachable
```

**Comment by nixpanic** [comment]
> /retest ci/centos/k8s-e2e-external-storage

#2036 has been merged, so hopefully this works now again

**Comment by nixpanic** [comment]
> /retest ci/centos/k8s-e2e-external-storage

Failed with #1969 :
```
Apr 29 13:50:47.670: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
```

**Comment by nixpanic** [comment]
@humblec @Madhu-1, I'd like to see this included in v3.4. Once Kubernetes v1.22 has been released, the capacity of BlockMode volumes can be reported. It does not hurt to have the functionality available, even when current Kubernetes versions do not collect the metrics.

**Comment by humblec** [comment]
> @humblec @Madhu-1, I'd like to see this included in v3.4. Once Kubernetes v1.22 has been released, the capacity of BlockMode volumes can be reported. It does not hurt to have the functionality available, even when current Kubernetes versions do not collect the metrics.

Sure, we can do that. As long as we have the confirmation here with the CI tests  which we already have with previous kube versions, we are good!

File: e2e/node.go [hunk 3]
```diff
@@ -42,3 +44,20 @@ func checkNodeHasLabel(c kubernetes.Interface, labelKey, labelValue string) erro
 	}
 	return nil
 }
+
+// List all nodes in the cluster (we have one), and return the IP-address.
+// Possibly need to add a selector, pick the node where a Pod is running?
```
Comment (humblec) [inline e2e/node.go:49]: Yeah, with a pod selector for E2E pods, we can actually do that and it would be good as well

File: e2e/pvc.go [hunk 4]
```diff
@@ -244,3 +245,44 @@ func checkPVSelectorValuesForPVC(f *framework.Framework, pvc *v1.PersistentVolum
 	}
 	return nil
 }
+
+func getMetricsForPVC(f *framework.Framework, pvc *v1.PersistentVolumeClaim, t int) error {
```
Comment (Madhu-1) [inline e2e/pvc.go:249]: as we randomly picking a node IP here will this work, if we don't have any pods running on this one? If not add a FIXME here?

File: e2e/utils.go [hunk 5]
```diff
@@ -299,6 +299,19 @@ func validateNormalUserPVCAccess(pvcPath string, f *framework.Framework) error {
 	if stdErr != "" {
 		return fmt.Errorf("failed to touch a file as non-root user %v", stdErr)
 	}
+
+	// metrics for BlockMode was added in Kubernetes 1.22
+	isBlockMode := false
+	if pvc.Spec.VolumeMode != nil {
+		isBlockMode = (*pvc.Spec.VolumeMode == v1.PersistentVolumeBlock)
+	}
+	if !isBlockMode || k8sVersionGreaterEquals(f.ClientSet, 1, 22) {
```
Comment (Madhu-1) [inline e2e/utils.go:311]: This should be for blockMode right? it should be `isBlockMode && k8sVersionGreaterEquals(f.ClientSet, 1, 22)` check?

File: internal/csi-common/nodeserver-default.go [hunk 6]
```diff
@@ -112,6 +114,53 @@ func (ns *DefaultNodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.No
...
+
+// blockNodeGetVolumeStats gets the metrics for a `volumeMode: Block` type of
+// volume. At the moment, only the size of the block-device can be returned, as
+// there are no secrets in the NodeGetVolumeStats request that enables us to
+// connect to the Ceph cluster.
+//
+// TODO: https://github.com/container-storage-interface/spec/issues/371#issuecomment-756834471
+func blockNodeGetVolumeStats(ctx context.Context, targetPath string) (*csi.NodeGetVolumeStatsResponse, error) {
+	args := []string{"--noheadings", "--bytes", "--output=SIZE", targetPath}
+	lsblkSize, _, err := util.ExecCommand(ctx, "/bin/lsblk", args...)
```
Comment (Madhu-1) [inline internal/csi-common/nodeserver-default.go]: validate stdErr also for the safer side?

**Review by Madhu-1 — COMMENTED** [review]
having few questions, otherwise LGTM

File: internal/csi-common/nodeserver-default.go [hunk 7]
```diff
@@ -112,6 +114,53 @@ func (ns *DefaultNodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.No
...
 	*/
 
+	stat, err := os.Stat(targetPath)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to get stat for targetpath %q: %v", targetPath, err)
+	}
+
+	if stat.Mode().IsDir() {
+		return filesystemNodeGetVolumeStats(ctx, targetPath)
+	} else if (stat.Mode() & os.ModeDevice) == os.ModeDevice {
```
Comment (Madhu-1) [inline internal/csi-common/nodeserver-default.go]: `NodeGetVolumeStats` common for both cephfs and rbd do we need to return an error If the driver is cephfs?

File: e2e/utils.go (hunk 5 above)
Comment (humblec) [reply e2e/utils.go:311]: Yes, it should be `&&` .

File: scripts/minikube.sh [hunk 8]
```diff
@@ -165,7 +165,7 @@ fi
 K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-"ExpandCSIVolumes=true"}
 
 #extra-config for kube https://minikube.sigs.k8s.io/docs/reference/configuration/kubernetes/
-EXTRA_CONFIG=${EXTRA_CONFIG:-"--extra-config=apiserver.enable-admission-plugins=PodSecurityPolicy"}
```
Comment (humblec) [inline scripts/minikube.sh:168]: @nixpanic this change is not related to this PR , Isnt it?

File: e2e/utils.go (hunk 5 above)
Comment (nixpanic) [reply e2e/utils.go:311]: No, even for non-BlockMode volumes we should test the metrics. That was not part of the e2e tests until now, this adds that as well.

Only from Kubernetes 1.22 the metrics for BlockMode PVCs are available.

File: internal/csi-common/nodeserver-default.go (hunk 7 above)
Comment (nixpanic) [reply internal/csi-common/nodeserver-default.go]: We can get the metrics for both CephFS and RBD. Maybe it is cleaner to split this out and have dedicated handlers for each node-plugin instead?

File: internal/csi-common/nodeserver-default.go (hunk 6 above)
Comment (nixpanic) [reply internal/csi-common/nodeserver-default.go]: We could, not sure if it is useful enough. The `err` should contain sufficient details already.

File: internal/csi-common/nodeserver-default.go (hunk 7 above)
Comment (Madhu-1) [reply internal/csi-common/nodeserver-default.go]: Sounds reasonable to split it out.

File: e2e/node.go (hunk 3 above)
Comment (nixpanic) [reply e2e/node.go:49]: that is an enhancement that may be needed later, it is not useful to add more complexity at this point

File: internal/csi-common/nodeserver-default.go (hunk 6 above)
Comment (Madhu-1) [reply internal/csi-common/nodeserver-default.go]: sometimes the err contains error codes but the stderr contains the human-readable error message.

File: e2e/utils.go (hunk 5 above)
Comment (Madhu-1) [reply e2e/utils.go:311]: okay, this means we are testing the filesystem mode pvc metrics not the block mode yet.

File: internal/csi-common/nodeserver-default.go (hunk 7 above)
Comment (nixpanic) [reply internal/csi-common/nodeserver-default.go]: This is done now, please check again.

File: e2e/utils.go (hunk 5 above)
Comment (nixpanic) [reply e2e/utils.go:311]: > okay, this means we are testing the filesystem mode pvc metrics not the block mode yet.

Yes, correct.

**Comment by Madhu-1** [comment]
@nixpanic can you please take a look at CI failure/?

File: internal/csi-common/utils.go [hunk 9]
```diff
@@ -185,3 +187,67 @@ func panicHandler(ctx context.Context, req interface{}, info *grpc.UnaryServerIn
 	}()
 	return handler(ctx, req)
 }
+
+func FilesystemNodeGetVolumeStats(ctx context.Context, targetPath string) (*csi.NodeGetVolumeStatsResponse, error) {
```
Comment (Madhu-1) [inline internal/csi-common/utils.go:195]: add a comment as function is exported?

**Review by Madhu-1 — APPROVED** [review]
LGTM, just had some minor comment

File: internal/rbd/nodeserver.go [hunk 10]
```diff
@@ -869,3 +869,58 @@ func (ns *NodeServer) xfsSupportsReflink() bool {
...
+	var err error
+	targetPath := req.GetVolumePath()
+	if targetPath == "" {
+		err = fmt.Errorf("targetpath %v is empty", targetPath)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	stat, err := os.Stat(targetPath)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to get stat for targetpath %q: %v", targetPath, err)
```
Comment (humblec) [inline internal/rbd/nodeserver.go:890]: `failed to get stats`..? (s/stat/stats)

Comment (humblec) [inline internal/rbd/nodeserver.go:890]: `failed to get stats`..? (s/stat/stats)

File: internal/cephfs/nodeserver.go [hunk 11]
```diff
@@ -325,3 +325,24 @@ func (ns *NodeServer) NodeGetCapabilities(ctx context.Context, req *csi.NodeGetC
...
+func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+	var err error
+	targetPath := req.GetVolumePath()
+	if targetPath == "" {
+		err = fmt.Errorf("targetpath %v is empty", targetPath)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	stat, err := os.Stat(targetPath)
+	if err != nil {
```
Comment (humblec) [inline internal/cephfs/nodeserver.go:339]: s/stat/stats?

File: e2e/node.go (hunk 3 above)
Comment (humblec) [reply e2e/node.go:49]: Sure.. it can be a followup item

File: internal/csi-common/utils.go (hunk 9 above)
Comment (nixpanic) [reply internal/csi-common/utils.go:195]: yes, will do!

File: internal/rbd/nodeserver.go (hunk 10 above)
Comment (nixpanic) [reply internal/rbd/nodeserver.go:890]: not really, it is `struct stat` from the `stat()` system call

**Comment by nixpanic** [comment]
> @nixpanic can you please take a look at CI failure/?

Ah, added the additional commit that uses the Kubernets e2e API to get metrics. This fails with:

    W0503 16:32:21.092662   80493 metrics_grabber.go:105] Did not receive an external client interface. Grabbing metrics from ClusterAutoscaler is disabled.

Will drop that commit now.

**Review by Madhu-1 — APPROVED** [review]
LGTM

**Comment by nixpanic** [comment]
@humblec and other concerns?

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
**ci/centos/mini-e2e-helm/k8s-1.20** and **ci/centos/mini-e2e/k8s-1.20** failed with #1969 

**ci/centos/upgrade-tests-cephfs** got hung when executing a `kubectl exec ... sha512sum`. Timeout happend after 90 minutes, but the test only has logs for 10 minutes (and then it hung).

**Comment by nixpanic** [comment]
**ci/centos/upgrade-tests-cephfs** failed again with #1969 

**ci/centos/mini-e2e/k8s-1.20** failed to bring up the minikube VM

**Review by humblec — APPROVED** [review]

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.20

Yet again #1969. But the job has been updated, so this is hopefully the last time :crossed_fingers:

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.20

Hit #1969 for cephfs testing.

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.20

Again failed due to #1969

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.20

So annoying, #1969  yet again... We need more debugging of the minikube VM (get libvirt status?) to find out the issue...

**Comment by nixpanic** [comment]
Both **ci/centos/mini-e2e/k8s-1.20** and **ci/centos/mini-e2e-helm/k8s-1.20** failed due to #1969

**Comment by Madhu-1** [comment]
@mergifyio rebase
